### PR TITLE
fix(barchartutils): condition check for categorydatasourceid

### PR DESCRIPTION
**Pull request by Ujjwal-UJ**
Closes Monitor - https://github.ibm.com/wiotp/Maximo-Asset-Monitor/issues/4717

**Summary**

- summary_here

**Change List (commits, features, bugs, etc)**

- [fix(barchartutils): condition check for categorydatasourceid rather than assigning undefined

**Acceptance Test (how to verify the PR)**

- tests here

**Regression Test (how to make sure this PR doesn't break old functionality)**

- tests here

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

File: packages/react/src/components/BarChartCard/barChartUtils.js [hunk 1]
```diff
@@ -241,7 +241,10 @@ export const formatChartData = (
         group.forEach((value) => {
           series.forEach((dataset) => {
             // if value is null, don't add it to the formatted chartData
-            if (!isNil(value[dataset.dataSourceId])) {
+            if (
+              !isNil(value[dataset.dataSourceId]) &&
+              (!isNil(value[categoryDataSourceId]) || timeDataSourceId)
```
Comment (herleraja) [inline packages/react/src/components/BarChartCard/barChartUtils.js:246]: may be ? 

```
(!isNil(value[categoryDataSourceId]) || !isNil(value[timeDataSourceId]))
```

Comment (davidicus) [reply packages/react/src/components/BarChartCard/barChartUtils.js:246]: What purpose does this second `!isNil(value[categoryDataSourceId])` in the ternary serve? This change is making this if statement look pretty complicated. Can we extract this out into a variable named in a descriptive way and use that in the if statement? ie. 
``` js
const isNilNonGroupedTimeDataSourceIdValue = timeDataSourceId && type !== BAR_CHART_TYPES.GROUPED ? !isNil(value[timeDataSourceId]) : true;
```

Comment (davidicus) [reply packages/react/src/components/BarChartCard/barChartUtils.js:246]: Just as an FYI, this ternary is creating a new "branch", that if not checked by some test will lower the code coverage of this file. These small changes with no test are what can lead to coverage issues that are seemingly unrelated down the road.

**Comment by davidicus** [comment]
Issues should outline how this bug can be reproduced in the PAL storybook if possible. The acceptance test part of this PR template should outline how we can check if this issues is resolved.

File: packages/react/src/components/BarChartCard/barChartUtils.js (hunk 1 above)
Comment (Ujjwal-UJ) [reply packages/react/src/components/BarChartCard/barChartUtils.js:246]: > What purpose does this second `!isNil(value[categoryDataSourceId])` in the ternary serve? This change is making this if statement look pretty complicated. Can we extract this out into a variable named in a descriptive way and use that in the if statement? ie.
> 
> ```js
> const nilCheckNonGroupedTimeDataSourceIdValue = timeDataSourceId && type !== BAR_CHART_TYPES.GROUPED ? !isNil(value[timeDataSourceId]) : true; // false if we want the check to fail without this
> ```

For Non-timeseries: if the data does not have value[categoryDataSourceId], it is assigning undefined and does not loads grouped bar chart at all. So `!isNil(value[categoryDataSourceId])` checks before assigning.
Also i will assign it to a variable as you suggested and add test cases too.

**Comment by davidicus** [comment]
@Ujjwal-UJ How is this PR coming? Are we still waiting on test cases? It looks as if the condition were were discussing has changed but I do not see any other commits?

**Comment by Ujjwal-UJ** [comment]
> @Ujjwal-UJ How is this PR coming? Are we still waiting on test cases? It looks as if the condition were were discussing has changed but I do not see any other commits?

@davidicus I was working on other issue, test cases are pending! will update that shortly

**Review by davidicus — APPROVED** [review]

**Pull request merged**